### ISSUE TITLE
[BUG] Proof composition fails when env::verify() is invoked twice

### ISSUE THREAD

**Issue by intoverflow** [risc0/risc0#1299]
## Bug Report

If the guest depends on more than 1 assumption, the prover fails with `invalid receipt format`.

## Steps to Reproduce

1. Checkout the branch `tcarstens/victor/composition-example`
2. Navigate to [`examples/composition`](https://github.com/risc0/risc0/tree/tcarstens/victor/composition-example/examples/composition)
3. Convince yourself that the [host](https://github.com/risc0/risc0/blob/tcarstens/victor/composition-example/examples/composition/src/main.rs#L29) and [guest](https://github.com/risc0/risc0/blob/tcarstens/victor/composition-example/examples/composition/methods/guest/src/main.rs#L26) are using the `add_assumption()` and `verify()` functions correctly, respectively.
4. Run the example in "dev mode" to convince yourself it works:

```
$ RUST_LOG=info RISC0_DEV_MODE=true cargo run --release
```

5. Run the example with proving enabled and note that the host fails when it calls `default_prover().prove()`:

```
$ RUST_LOG=info RISC0_DEV_MODE=false cargo run --release
```

## Expected behavior

Expected a receipt to be generated with no outstanding assumptions.

## Your Environment

- risc0-zkvm version: `cargo-risczero 0.21.0-alpha.1` (installed from this branch using `cargo install --path risc0/cargo-risczero`)
- Rust version: `rustc 1.73.0 (cc66ad468 2023-10-03)`
- Platform/OS: `Linux pop-os 5.19.0-76051900-generic #202207312230~1663791054~22.04~28340d4 SMP PREEMPT_DYNAMIC Wed S x86_64 x86_64 x86_64 GNU/Linux`

## Additional context

This issue has also been observed in zkVM `v0.20.0-rc.1`.

**Comment by intoverflow** [comment]
I wonder if this TODO is related: https://github.com/risc0/risc0/blob/main/risc0/zkvm/src/host/recursion/receipt.rs#L101

**Comment by nategraf** [comment]
Ok, I've fixed this in a867e0db7850c30f3ee5209aba2b854aa8ec8611. The issue is that the assumptions list as tracked by the executor was in reverse order compared to how it is tracked assembled by the guest. I'll make sure to add a test that checks functionality with two or more assumptions, as this is a clear gap.
